### ISSUE TITLE
Rule proposal: prefer-optional-catch-binding

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#10434]
**Please describe what the rule should do:**

This rule reports and fixes unused `catch` bindings.

```js
try {
    f()
} catch (err) { // unused!
    g()
}
```

We can modify such code safely to simple:

```js
try {
    f()
} catch {
    g()
}
```

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
//✗ BAD
try {
    f()
} catch (err) { // unused!
    g()
}

//✓ GOOD
try {
    f()
} catch (err) {
    g(err)
}

try {
    f()
} catch {
    g()
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This rule will encourage to use new syntax in ES2019.
It makes more readable to remove unused variables.
The autofix is useful.

**Comment by ljharb** [comment]
This is awesome.

**Comment by not-an-aardvark** [comment]
Aside from the autofix, is there any difference between this rule and the `no-unused-vars` option for catch bindings?

**Comment by mysticatea** [comment]
The `no-unused-vars` has many options. Especially,

- it can ignore `catch` bindings by `caughtErrors` option.
- it can ignore `catch` bindings by `varsIgnorePattern` option.

For example, I'm using `varsIgnorePattern` option to ignore unused variables in parameters or `catch`. This rule will work in such case.

Hmm, but indeed, significant difference might be only autofix.

**Comment by ljharb** [comment]
Another difference is the explicit presence of a more precise error message and documentation.

**Comment by not-an-aardvark** [comment]
¯\\\_(ツ)_/¯ I'm not sure the difference is very substantial, in both cases the solution would be to either use the variable or remove it. The difference is just that developers previously weren't allowed to remove the variable, and now they are.

**Comment by ljharb** [comment]
I'd love to see this reopened and revisited.
